### ISSUE TITLE
[HOLD for payment 2022-05-20][$1000] Feature Request: Show link when hover over hyperlink

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7844]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to chat 
2. Send any text hyperlinked  
3. Hover over or hard press over link 

## Expected Result:
Some tool tip displayed so that, when the hyperlink is hovered over, the destination link shows. 

## Actual Result:
Nothing happens on hover 

## Problem: 
Currently users can use markdown to add hyperlinks to chats.  The person who receives the link can't easily view the link to know where goes (inc. potential phishing links)

## Solution: 
Add a tool tip so that, when the hyperlink is hovered over, the destination link shows.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.39-1 
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
<img width="293" alt="image - 2022-02-20T011335 541" src="https://user-images.githubusercontent.com/43995119/154831033-70c5d3ba-a835-41ae-8ceb-f795bb93a182.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1644366109776009

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by sobitneupane** [comment]
**PROPOSAL**
To achieve this we need to make some changes in src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js
https://github.com/Expensify/App/blob/d9355db0e9685021c3be1debca3c43751ad2d759/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js#L54-L70

```
<Tooltip text={attrHref}>
    <AnchorForCommentsOnly
        href={attrHref}
        isAuthTokenRequired={isAttachment}

        // Unless otherwise specified open all links in
        // a new window. On Desktop this means that we will
        // skip the default Save As... download prompt
        // and defer to whatever browser the user has.
        // eslint-disable-next-line react/jsx-props-no-multi-spaces
        target={htmlAttribs.target || '_blank'}
        rel={htmlAttribs.rel || 'noopener noreferrer'}
        style={{...props.style, ...parentStyle}}
        key={props.key}
        fileName={fileName}
    >
        <TNodeChildrenRenderer tnode={props.tnode} />
    </AnchorForCommentsOnly>
</Tooltip>
```
**Output:**
<img width="1440" alt="Screen Shot 2022-02-20 at 14 53 40" src="https://user-images.githubusercontent.com/25876548/154835781-bba3ef5f-83a1-49cf-bc18-cc6ae6047e59.png">

**Comment by parasharrajat** [comment]
This is an extra feature. it could be good to have but I don't it is needed at this stage.

**Comment by kadiealexander** [comment]
Agree, but if it's an easy fix I say we go for it. @mallenexpensify is right that it poses a small security risk to click on a link when you're not sure of the destination.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sketchydroide (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by parasharrajat** [comment]
Then I don't think a tooltip is best suited bfor this. Like other websites, hovering over the link the browser should show the link address on the cornor of the browser.

But this is not happening in our app so we should do that even though we are planning to show tooltip.

**Comment by sketchydroide** [comment]
>Then I don't think a tooltip is best suited bfor this. Like other websites, hovering over the link the browser should show the link address on the cornor of the browser.
>
>But this is not happening in our app so we should do that even though we are planning to show tooltip.

I think for now let's show the tooltip, this is what slack seems to do, and for the rest of the none web based platforms I think it makes more sense.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by sketchydroide** [comment]
😄 @mallenexpensify you have been chosen I was goint to tag you, but seems like it is no longer necessary

**Comment by parasharrajat** [comment]
But still Tooltips are only web thing.

**Comment by mallenexpensify** [comment]
Full circle!  @parasharrajat is there another term we should use to denote what we're calling/considering a tooltip on Desktop? 
![image](https://user-images.githubusercontent.com/22508304/155026604-50df8682-5245-47ad-a671-9f9b06c41518.png)

We definitely want a mobile/app option too, right now there is no way to know where a hyperlink will send you.  
![IMG_5DAA83E20312-1](https://user-images.githubusercontent.com/22508304/155026912-318da127-7e4c-44d4-83af-750d6f96fa8c.jpeg)

How about, when someone taps then holds any link (hyperlinks and other domain links) we show the destination link  above or below 'copy URL to clipboard`? Reasoning is because it's possible to use markdown to have a visible link go to a different site
<img width="367" alt="image" src="https://user-images.githubusercontent.com/22508304/155027600-fb045965-5ac6-49d3-bfb3-aef48a0318ad.png">

Alt option would be to never allow links to go to other sites.

**Comment by parasharrajat** [comment]
Sorry for the confusion @mallenexpensify. Tooltip works on both web and desktop but not on the native apps (Android|IOS).

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mallenexpensify** [comment]
Job posted - https://www.upwork.com/jobs/~01f57f9d1472302ddc
@tgolen , can you review @sobitneupane 's proposal above https://github.com/Expensify/App/issues/7844#issuecomment-1046194797

Also, for the [lower part of this comment](https://github.com/Expensify/App/issues/7844#issuecomment-1047237439), do you think we should address that here?

**Comment by parasharrajat** [comment]
@sobitneupane Can we please drill down in the `AnchorForCommentsOnly` to show the tooltip? `AnchorForCommentsOnly` is multipurpose. The tooltip should be added to the closest native node.

@mallenexpensify I agree, we if we want to show the URL on mobile, we can show that over the Copy to clipboard context item in the popup menu.

**Comment by mallenexpensify** [comment]
> @tgolen , can you review @sobitneupane 's proposal above https://github.com/Expensify/App/issues/7844#issuecomment-1046194797

sorry @parasharrajat , forgot you were assigned as C+ :ohnothing:   Let's see if we can tackle both mobile and web/desktop via this issue

**Comment by parasharrajat** [comment]
Do you prefer the tooltip on Mobile or fine with showing the link on the popup menu ?

**Comment by sobitneupane** [comment]
> @sobitneupane Can we please drill down in the `AnchorForCommentsOnly` to show the tooltip? `AnchorForCommentsOnly` is multipurpose. The tooltip should be added to the closest native node.

If we want to show tooltip both in desktop and web app, we should add `<Tooltip>` around
https://github.com/Expensify/App/blob/87230549a3d9c299f7abbc98012f4143383c2c19/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.js#L72-L85

If we want to show tooltip only in desktop, we should add a new file index.desktop.js with the codes in index.js with above change.

Showing url in popover can be quite tricky. Length of url (even only domain) can be an issue.

**Comment by parasharrajat** [comment]
Do you see any other way to show the URL on Mobile @sobitneupane ?  Holding proposal review until we finalize the changes that need to be done for
1. Native?
2. Web & Desktop ( Tooltip)

**Comment by parasharrajat** [comment]
Bump questions :arrow_up:

**Comment by parasharrajat** [comment]
@mallenexpensify https://github.com/Expensify/App/issues/7844#issuecomment-1048975682

**Comment by mallenexpensify** [comment]
> Do you prefer the tooltip on Mobile or fine with showing the link on the popup menu? But the link could be very big and we may have to trim it. But I think most important part is the domain so it should be properly visible.

I think we'd use iOS and Android defaults, if there is such a thing.  ie. I think long-pressing the link would open the popup from the bottom on iOS where you'd be able to see the link at the top of the pop up then have options below that for sharing/copying.  Not sure about Android

**Comment by parasharrajat** [comment]
Android will have the same effect. Got it, Thanks.

@sobitneupane Could you please share an updated proposal with the new requirements?

**Comment by mallenexpensify** [comment]
Does _everyone_ agree that what I proposed is the best? I'm not an expert in UI/UX...

**Comment by tgolen** [comment]
I'm OK with that proposal given the feature request. Tooltips are hard cross-platform and so I think trying to stick close to native platform functionality is OK. 

However, I can't shake the feeling it just feels like this feature a bit much and I would actually vote that I don't want the feature because I don't see the problem (or haven't experienced it).

**Comment by mallenexpensify** [comment]
I think that not being able to know the destination of links your clicking is a security issue. Phishing 101 is to spoof an email and link to get people to click it.  Isn't that the same, or similar, to what we're wanting to fix/update here?

**Comment by tgolen** [comment]
Yeah, you're right. It makes sense from a security perspective. I forgot about that.

**Comment by tgolen** [comment]
@sobitneupane I think we are still waiting on this from you:

> @sobitneupane Could you please share an updated proposal with the new requirements?

**Comment by sketchydroide** [comment]
@sobitneupane any updates?

**Comment by mallenexpensify** [comment]
Doubled price (should have done a while ago...)
https://www.upwork.com/jobs/~018bbb25aad73a12f4

**Comment by mallenexpensify** [comment]
Doubled price to $1000
https://www.upwork.com/jobs/~018bbb25aad73a12f4

**Comment by rushatgabhane** [comment]
>  Since there is no hover on mobile, show the link URL in the context menu for a long press.

Just dropping in a possible solution for mobile, from [slack.](https://expensify.slack.com/archives/C01GTK53T8Q/p1649883736443399?thread_ts=1649883736.443399&cid=C01GTK53T8Q)

**Comment by eVoloshchak** [comment]
> > Since there is no hover on mobile, show the link URL in the context menu for a long press.
> 
> Just dropping in a possible solution for mobile, from [slack.](https://expensify.slack.com/archives/C01GTK53T8Q/p1649883736443399?thread_ts=1649883736.443399&cid=C01GTK53T8Q)

**Proposal for native:**
Making use of the [MenuItem's](https://github.com/Expensify/App/blob/main/src/components/MenuItem.js) `description` prop, the result looks like this:


https://user-images.githubusercontent.com/9059945/163721970-b29c5b98-3d53-4b40-a8cd-18b21d2797dd.mp4



**Implementation:**

1. Pass an additional `description` prop through [ReportActionContextMenu](https://github.com/Expensify/App/blob/main/src/pages/home/report/ContextMenu/ReportActionContextMenu.js) -> [PopoverReportActionContextMenu](https://github.com/Expensify/App/blob/main/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js) -> [BaseReportActionContextMenu](https://github.com/Expensify/App/blob/main/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js) -> [ContextMenuItem](https://github.com/Expensify/App/blob/main/src/components/ContextMenuItem.js) -> to the  [MenuItem](https://github.com/Expensify/App/blob/main/src/components/MenuItem.js)
2. In [BaseAnchorForCommentsOnly/index.native.js](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.native.js#L66) pass the link (`displayName`) to `ReportActionContextMenu.showContextMenu` and check that it's not an email:
```javascript
ReportActionContextMenu.showContextMenu(
    Str.isValidEmail(this.props.displayName) ? ContextMenuActions.CONTEXT_MENU_TYPES.EMAIL : ContextMenuActions.CONTEXT_MENU_TYPES.LINK,
    event,
    this.props.href,
    lodashGet(linkRef, 'current'),
    null,
    null,
    null,
    () => {},
    () => {},
    !Str.isValidEmail(this.props.displayName) && this.props.displayName
);
```
Im guessing the same should also be implemented for mWeb?

**Comment by parasharrajat** [comment]
We don't need to pass the additional data from actions. The link value is available in the selection already.

Rest seems fine to fine. 

@Expensify/design Does the video [here](https://github.com/Expensify/App/issues/7844#issuecomment-1100904641) look good?

**Comment by shawnborton** [comment]
Looks good. Can we see an example of the desktop tooltip version with an incredibly long link?

**Comment by eVoloshchak** [comment]
> Looks good. Can we see an example of the desktop tooltip version with an incredibly long link?

https://user-images.githubusercontent.com/9059945/163836319-f0bece44-a669-4445-8602-706704997154.mp4

I'm guessing the Tooltip component should me modified.
How should it look like?

1. We could limit the length of the link
2. We could add `maxWidth: Dimensions.get('window').width` to the [tooltipWrapperStyle](https://github.com/Expensify/App/blob/6c53be1ea7a022e8a5e788cc7e7efd62c4573474/src/styles/getTooltipStyles.js#L105), which would look like this:

https://user-images.githubusercontent.com/9059945/163852410-e1370b57-5f57-416c-b852-c8f8e1d609e9.mp4


But as you can see, pointer is calculated to be at the center of the link, which looks odd, because it is not at the center of the tooltip.
3. Possibly another approach (limit tooltip only to the left part of the screen, the one with chat)?

And what should we do with the native tooltip in browsers? The one that appears at the bottom left when you hover over a link

**Comment by shawnborton** [comment]
I think we can ignore the native tooltip and just let it be. 

I agree that we want some kind of max-width on the tooltip itself. Perhaps if the max-width is just the total width of the message width then the pointer will always be center-aligned.

**Comment by parasharrajat** [comment]
There is another issue which is setting the tooltip maxWidth so we don't have to worry about it here. 

@eVoloshchak 's proposal looks good to me. The proposal discussion starts from [here](https://github.com/Expensify/App/issues/7844#issuecomment-1100904641).

cc: @tgolen 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by mallenexpensify** [comment]
@tgolen bumping to daily, @eVoloshchak 's proposal needs review

**Comment by tgolen** [comment]
Ah, sorry I missed this one. The proposal looks OK. 🟢  to hire @eVoloshchak

**Comment by mallenexpensify** [comment]
Hired @eVoloshchak and assigned to GH!

**Comment by eVoloshchak** [comment]
@parasharrajat 
I've encountered a problem on Android (haven't tested on iOS yet, but pretty sure it is present there too). 
When we have a message with text and a link, long pressing on text will pop up the context menu for the message (which is correct), but long pressing on the link will also pop up the context menu for the message (which is not correct, it should be a context message with **Copy URL to clipboard**). While if you have a message with just a link, It works correctly.
Here's a video to illustrate

https://user-images.githubusercontent.com/9059945/165731370-3611fef1-7fac-4802-8afa-83da6b37501c.mp4

The issue occurs because if [PressableWithSecondaryInteraction](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.native.js) is inline, we use Text instead of Pressable and it somehow interferes with [LongPressGestureHandler](https://github.com/Expensify/App/blob/9bf91283eb9ac89409154bd1f45c94aa44530906/src/components/PressableWithSecondaryInteraction/index.native.js#L19)
```javascript
const Node = props.inline ? Text : Pressable;
```
I tried to fix this and found out, that if we completely remove LongPressGestureHandler and add onLongPress handler to Node, it works just fine

https://user-images.githubusercontent.com/9059945/165732994-886ed14f-da26-409b-9bba-0a0a689dcde9.mp4

So [this](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.native.js) file would look like

```javascript
const PressableWithSecondaryInteraction = (props) => {
    // Use Text node for inline mode to prevent content overflow.
    const Node = props.inline ? Text : Pressable;
    return (
        <Node
            ref={props.forwardedRef}
            onPress={props.onPress}
            onPressIn={props.onPressIn}
            onPressOut={props.onPressOut}
            onLongPress={(e) => {
                e.preventDefault();
                HapticFeedback.trigger();
                props.onSecondaryInteraction(e);
            }}
            {...props}
        >
            {props.children}
        </Node>
    );
};
```
Since you were the author of the [PR](https://github.com/Expensify/App/pull/6392) that introduced this change, I thought you might know more.
Is approach I proposed acceptable?
Do we need LongPressGestureHandler?
I poked around the app and it doesn't seem to break anything, but maybe you have more context
Thanks!

**Comment by parasharrajat** [comment]
Hey, @eVoloshchak Thanks for looking into it. It is a different issue than the one we are trying to solve here. So you don't have to fix that in this PR. 

we need to use `Text` for inline mode and we also need the `LongPressGestureHandler` as it fixes one of the old issues. 

but we have a separate logged issue for the problem you mentioned and if you want to solve it, head over to https://github.com/Expensify/App/issues/8311.

**Comment by eVoloshchak** [comment]
> Hey, @eVoloshchak Thanks for looking into it. It is a different issue than the one we are trying to solve here. So you don't have to fix that in this PR.

Thanks, go it.

I'm ready to submit the PR, but it's gonna be impossible for the QA to test it for `mWeb` due to https://github.com/Expensify/App/issues/8311.
Should I submit it or wait for this issue to be resolved?

**Comment by parasharrajat** [comment]
PR under review

**Comment by sketchydroide** [comment]
no update

**Comment by trjExpensify** [comment]
👋 looks like this didn't quite work completely, issue reported with mWeb: https://github.com/Expensify/App/issues/8960#issue-1233333508

**Comment by eVoloshchak** [comment]
Issue isn't caused by this PR, more details [here](https://github.com/Expensify/App/issues/8960#issuecomment-1129233056).

**Comment by mallenexpensify** [comment]
Paid @eVoloshchak , it's working on web and desktop (Version 1.1.64-0). No comments on [the PR](https://github.com/Expensify/App/pull/8815) the past week. 

@parasharrajat , hired you for C+ payment, can you accept the offer?

<img width="322" alt="image" src="https://user-images.githubusercontent.com/22508304/169591301-c702cdc6-ca19-4996-918d-e632c9cdeb2e.png">

<img width="457" alt="image" src="https://user-images.githubusercontent.com/22508304/169591309-a23fa292-8345-4f22-8f1b-d0d15070b30a.png">

**Comment by parasharrajat** [comment]
@mallenexpensify Done.

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat , paid $1000 for C+.
